include/EncoderDecoder/AbstractEncoderDecoder.h

```diff
@@ -0,0 +1,57 @@
+#ifndef ABSTRACT_ENCODER_DECODER_H
+#define ABSTRACT_ENCODER_DECODER_H
+
+#include <string>
+
+template <typename MappingType, typename EncodedType>
+class AbstractEncoderDecoder {
+public:
+	/**
+	 * @brief Encodes a given plaintext string.
+	 *
+	 * @param plaintext String to encode
+	 * @return std::pair<bool, EncodedType> A pair representing whether the
+	 * encoding succeeded and the result of the encoding operation. If the first
+	 * value is false, the value of the second is undefined.
+	 */
+	virtual std::pair<bool, EncodedType> encode(
```

Comment:
Why return a pair of `bool` and value instead of returning a value and throwing on error?

Reply:
I wasn't sure about that. I also think in retrospect that throwing may be more appropriate, since there are several failure cases.
This is a non-trivial change (that is relevant in several places), so I opened an issue for it (#39).